### DISCUSSION TITLE
User Auto-Provisioning and Ident

### DISCUSSION THREAD

**Discussion by man-github** [SoftEtherVPN/SoftEtherVPN#1766] [Polls]
Applications setup in the cloud, hosted on servers behind a VPN server like SoftEther, prevents unauthorized access, making VPN so invaluable for security. 

For large enterprises with thousands of users scattered world-wide, hired by regional offices, that need multiple VPN servers geo-strategically located, SoftEther presents some very key challenges in adoption.

1. SoftEther requires a lot of manual effort for user provisioning.
2. Applications being securely accessed require another round of credential validation to identify the user.
3. The logging statements of SoftEther make it is very difficult to co-relate the VPN authenticated user from the IP address witnessed by the application.  

SSL Client Certificate based authentication is surely more reliable than traditional username/password authentication.
The Client certificates can be automatically pushed to each end-point using various mechanisms, including Microsoft Group Policy. Fraudulent use can be controlled by keeping such certificates short-lived, or CRL, or tens of other creative mechanisms.

SoftEther's support for SSTP makes it super attractive for enterprises that seek to "tether" company-owned end-points from misuse.
Most importantly it means end-points do not need an additional VPN client software to join the party, and can enjoy a robust native mechanism.

However adoption of SoftEther would be so much easier, if:

1. Users authenticated via SSL Certificate authentication could be auto-provisioned (email / CN in the SSL Certificate).
2. SoftEther executes a script like OpenVPN with username & DHCP allocated IP address as environement variables, to push the IP:User translation into an external / remote database. Thus any application, secured by SoftEther would easily gain user-awareness by ident'ing the user from such database.
3. The logging statements should be machine-readable, hopefully SIEM friendly.

With the above features, the scalability of SoftEther would jump manifold particularly if one wanted to setup say a few hundered VPN servers across various data centers on the Internet. This would also make monitoring application usage so easy, for security purposes.

**Comment by shakibamoshiri** [comment]
You are right to some degree.
The problem is that this project it huge and has mulit parts but does not have enough developers to progress all part accordingly 
For example it has RPC-JSON API which makes it easy to manage the server remotely but it is not complete  

Just as an example of what you mentioned "SoftEther requires a lot of manual effort for user provisioning." you can simply add any number of users, anywhere in minutes using RPC-JSON + a client 
sample: https://github.com/shakibamoshiri/secli

for server provisioning you can use a remote runner to have a fully setup in minutes anywhere you like.
finally SE is not a "plug-and-play" product, so for enterprise usage, you have to
- read RPC-JSON API Doc
-  learn "vpncmd"

And more donation and sponsoring help SE better.

**Reply by man-github** [reply]
Followed your guidance, but not sure if I understood the content in the links.

Kindly enlighten a bit more, here's what I need specifically:
My web service is running in cloud and secured behind the Softether VPN service, that authenticates clients via signed SSL certificates.
My web service needs to identify the username (or some other parameter in the client's SSL certificate) from the client IP address it witnesses.
My web service can execute HTTP(s)/REST requests for this.
_Can you please share a curl command that i may consider as a template / example?_

Donating or sponsoring SE is definitely a great idea.
Unfortunately the folks at SE are too busy to reply on emails or even on the forums.
And that makes it very difficult to build a winning case for them.

_Ps. don't mind the delayed response, was unwell._

**Reply by shakibamoshiri** [reply]
if you can GET/POST http request and you know a programming language then check out their API
[SoftEther VPN Server JSON-RPC API Suite Document](https://github.com/SoftEtherVPN/SoftEtherVPN/tree/master/developer_tools/vpnserver-jsonrpc-clients)

here is an example to see list of a HUB users 

```bash
#!/bin/bash

# hub name => VPN
curl_data='{"jsonrpc":"2.0","id":"rpc_call_id","method":"EnumUser","params":{"HubName_str":"VPN"}}';

# server credential 
server_pass='**************';
server_addr=vpn.example.com
server_port=1234;

# query 
curl -sL --insecure  -X POST -H "X-VPNADMIN-PASSWORD: ${server_pass}"  https://${server_addr}:${server_port}/api/ -d "${curl_data}" | jq '.result.UserList'
```
